addons/hr_contract/report/hr_contract_history.py

```diff
@@ -0,0 +1,115 @@
+    def _compute_display_name(self):
+        for contract in self:
+            contract.display_name = '%s %s' % (contract.employee_id.name, _("'s Contracts History"))
+
+    @api.model
+    def _get_fields(self):
+        return ','.join('contract.%s' % name for name, field in self._fields.items()
+                        if field.store 
+                        and field.type not in ['many2many', 'one2many', 'related']
+                        and field.name not in ['id', 'contract_id', 'employee_id', 'date_hired', 'is_under_contract'])
+
+    def init(self):
+        tools.drop_view_if_exists(self.env.cr, self._table)
+        # Reference contract is the one with the latest start_date.
+        self.env.cr.execute("""CREATE or REPLACE VIEW %s AS (
+            WITH contract_information AS (
+                SELECT DISTINCT employee_id,
+                                FIRST_VALUE(id) OVER w_partition AS id,
+                                MAX(CASE WHEN state='open' THEN 1 ELSE 0 END) OVER w_partition AS is_under_contract
+                FROM   hr_contract AS contract
+                WHERE  contract.state <> 'cancel' AND contract.active = true
+                WINDOW w_partition AS (
+                    PARTITION BY contract.employee_id
+                    ORDER BY     contract.date_start DESC
+                    RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
+                )
+            )
+            SELECT     employee.id AS id,
+                       employee.id AS employee_id,
+                       contract.id AS contract_id,
+                       contract_information.is_under_contract::bool AS is_under_contract,
+                       employee.first_contract_date AS date_hired,
+                       %s
+            FROM       hr_contract AS contract
+            INNER JOIN contract_information ON contract.id = contract_information.id
+            RIGHT JOIN hr_employee AS employee ON contract_information.employee_id = employee.id
+            WHERE      employee.active = true OR contract.state='draft'
+        );""" % (self._table, self._get_fields()))
+
+    @api.depends('employee_id.contract_ids')
+    def _compute_contract_ids(self):
+        # Optimising number of queries in case of multi
+        contracts = self.env['hr.contract'].search(
+            ['&',
+             ('state', '!=', 'cancel'),
+             ('id', 'in', [contract.id for employee in self.mapped('employee_id') for contract in employee.contract_ids])],
+            order='date_start desc'
+        )
+
+        contract_history_dic = {contract_history.id: contract_history for contract_history in self}
+        for contract in contracts:
+            contract_history_dic[contract.employee_id.id].contract_ids += contract
+
+        if len(contracts) < len(self.ids):
+            for employee_id in (set(self.ids) - set(contracts.mapped('employee_id').mapped('id'))):
+                contract_history_dic[employee_id].contract_ids = ()
```

Comment:
This is quite difficult to read :)

What do you think about:
``` python3
sorted_contracts = self.mapped('employee_id.contract_ids').sorted('date_start', reverse=True)

mapped_employee_contracts = defaultdict(lambda: self.env['hr.contract'])
for contract in sorted_contracts:
    if contract.state != 'cancel':
        mapped_employee_contracts[contract.employee_id] |= contract

for history in self:
    history.contract_ids = mapped_employee_contracts[history.employee_id.id]
```

Reply:
👍 This is where I see that I still have to learn Python 🙂